File: frontend/src/api/recipes.ts
import { del, get, post, request } from "./client";

export const UNIT_OPTIONS = [
  "unit",
  "ml",
  "l",
  "litro",
  "g",
  "mg",
  "kg",
  "cucharada",
  "cucharadita",
  "cuchara",
  "taza",
  "pizca",
  "onza",
  "lb",
  "paquete",
  "lata",
  "botella",
] as const;

export type RecipeUnit = (typeof UNIT_OPTIONS)[number];

export type Category = {
  id: string;
  name: string;
  color: string;
};

export type PriceObservation = {
  id: string;
  ingredient_id: string;
  price: number;
  quantity: number;
  unit: RecipeUnit;
  source?: string | null;
  observed_at: string;
};

export type Ingredient = {
  id: string;
  name: string;
  default_unit?: RecipeUnit | null;
  notes?: string | null;
  latest_price?: PriceObservation | null;
};

export type RecipeIngredient = {
  id?: string;
  ingredient_id?: string | null;
  name: string;
  quantity?: number | null;
  unit?: RecipeUnit | null;
  note?: string | null;
  position: number;
  estimated_cost?: number | null;
};

export type RecipeStep = {
  id?: string;
  text: string;
  position: number;
};

export type RecipeSummary = {
  id: string;
  title: string;
  description?: string | null;
  category_id?: string | null;
  category_name?: string | null;
  favorite: boolean;
  rating?: number | null;
  servings: number;
  image_ref?: string | null;
  estimated_cost?: number | null;
  estimated_cost_per_serving?: number | null;
  created_at: string;
  updated_at: string;
};

export type Recipe = RecipeSummary & {
  user_note?: string | null;
  prep_minutes?: number | null;
  cook_minutes?: number | null;
  ingredients: RecipeIngredient[];
  steps: RecipeStep[];
};

export type RecipePayload = {
  title: string;
  description?: string | null;
  category_id?: string | null;
  favorite: boolean;
  rating?: number | null;
  user_note?: string | null;
  servings: number;
  prep_minutes?: number | null;
  cook_minutes?: number | null;
  image_ref?: string | null;
  ingredients: Omit<RecipeIngredient, "id" | "estimated_cost">[];
  steps: Omit<RecipeStep, "id">[];
};

export type MealPlanEntry = {
  id: string;
  weekday: number;
  position: number;
  recipe_id: string;
  note?: string | null;
  recipe: RecipeSummary;
};

export type MealPlanEntryPayload = {
  weekday: number;
  position: number;
  recipe_id: string;
  note?: string | null;
};

export function listRecipes(params: {
  q?: string;
  favorite?: boolean;
  category_id?: string;
}) {
  const query = new URLSearchParams();
  if (params.q) query.set("q", params.q);
  if (params.favorite !== undefined) query.set("favorite", String(params.favorite));
  if (params.category_id) query.set("category_id", params.category_id);
  const suffix = query.toString() ? `?${query.toString()}` : "";
  return get<RecipeSummary[]>(`/api/recipes${suffix}`);
}

export function getRecipe(id: string) {
  return get<Recipe>(`/api/recipes/${id}`);
}

export function createRecipe(payload: RecipePayload) {
  return post<Recipe>("/api/recipes", payload);
}

export function updateRecipe(id: string, payload: RecipePayload) {
  return request<Recipe>(`/api/recipes/${id}`, { method: "PUT", body: payload });
}

export function deleteRecipe(id: string) {
  return del<{ status: string }>(`/api/recipes/${id}`);
}

export function listCategories() {
  return get<Category[]>("/api/categories");
}

export function createCategory(payload: { name: string; color: string }) {
  return post<Category>("/api/categories", payload);
}

export function updateCategory(id: string, payload: { name: string; color: string }) {
  return request<Category>(`/api/categories/${id}`, { method: "PUT", body: payload });
}

export function listIngredients() {
  return get<Ingredient[]>("/api/ingredients");
}

export function createIngredient(payload: {
  name: string;
  default_unit?: RecipeUnit | null;
  notes?: string | null;
}) {
  return post<Ingredient>("/api/ingredients", payload);
}

export function updateIngredient(
  id: string,
  payload: { name: string; default_unit?: RecipeUnit | null; notes?: string | null },
) {
  return request<Ingredient>(`/api/ingredients/${id}`, { method: "PUT", body: payload });
}

export function addIngredientPrice(
  ingredientId: string,
  payload: { price: number; quantity: number; unit: RecipeUnit; source?: string | null },
) {
  return post<PriceObservation>(`/api/ingredients/${ingredientId}/prices`, payload);
}

export function listMealPlan() {
  return get<MealPlanEntry[]>("/api/meal-plan");
}

export function saveMealPlan(payload: MealPlanEntryPayload[]) {
  return request<MealPlanEntry[]>("/api/meal-plan", { method: "PUT", body: payload });
}

File: backend/src/app/routes/recipes.py
from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlmodel import Session, delete, select

from app.database import get_session
from app.models import Recipe, RecipeCategory, RecipeIngredient, RecipeStep, utcnow
from app.schemas import (
    RecipeIngredientRead,
    RecipeRead,
    RecipeStepRead,
    RecipeSummary,
    RecipeWrite,
)
from app.services.pricing import estimated_ingredient_cost

router = APIRouter(prefix="/api/recipes", tags=["recipes"])


def _category_name(session: Session, category_id: str | None) -> str | None:
    if not category_id:
        return None
    category = session.get(RecipeCategory, category_id)
    return category.name if category else None


def _ingredient_reads(
    session: Session, recipe_id: str
) -> tuple[list[RecipeIngredientRead], float | None]:
    rows = session.exec(
        select(RecipeIngredient)
        .where(RecipeIngredient.recipe_id == recipe_id)
        .order_by(RecipeIngredient.position)
    ).all()
    if not rows:
        return [], None

    items: list[RecipeIngredientRead] = []
    known_total = 0.0
    has_unknown = False
    for row in rows:
        cost = estimated_ingredient_cost(session, row)
        if cost is None:
            has_unknown = True
        else:
            known_total += cost
        items.append(
            RecipeIngredientRead(
                id=row.id,
                ingredient_id=row.ingredient_id,
                name=row.name,
                quantity=row.quantity,
                unit=row.unit,
                note=row.note,
                position=row.position,
                estimated_cost=cost,
            )
        )
    return items, None if has_unknown and known_total == 0 else round(known_total, 2)


def _step_reads(session: Session, recipe_id: str) -> list[RecipeStepRead]:
    rows = session.exec(
        select(RecipeStep)
        .where(RecipeStep.recipe_id == recipe_id)
        .order_by(RecipeStep.position)
    ).all()
    return [RecipeStepRead(id=row.id, text=row.text, position=row.position) for row in rows]


def _summary(session: Session, recipe: Recipe) -> RecipeSummary:
    _, cost = _ingredient_reads(session, recipe.id)
    return RecipeSummary(
        id=recipe.id,
        title=recipe.title,
        description=recipe.description,
        category_id=recipe.category_id,
        category_name=_category_name(session, recipe.category_id),
        favorite=recipe.favorite,
        rating=recipe.rating,
        servings=recipe.servings,
        image_ref=recipe.image_ref,
        estimated_cost=cost,
        estimated_cost_per_serving=round(cost / recipe.servings, 2) if cost else None,
        created_at=recipe.created_at,
        updated_at=recipe.updated_at,
    )


def _read(session: Session, recipe: Recipe) -> RecipeRead:
    summary = _summary(session, recipe)
    return RecipeRead(
        **summary.model_dump(),
        user_note=recipe.user_note,
        prep_minutes=recipe.prep_minutes,
        cook_minutes=recipe.cook_minutes,
        ingredients=_ingredient_reads(session, recipe.id)[0],
        steps=_step_reads(session, recipe.id),
    )


def _validate_category(session: Session, category_id: str | None) -> None:
    if category_id and not session.get(RecipeCategory, category_id):
        raise HTTPException(status_code=404, detail="Category not found")


def _replace_children(session: Session, recipe: Recipe, payload: RecipeWrite) -> None:
    session.exec(delete(RecipeIngredient).where(RecipeIngredient.recipe_id == recipe.id))
    session.exec(delete(RecipeStep).where(RecipeStep.recipe_id == recipe.id))
    for index, item in enumerate(payload.ingredients):
        session.add(
            RecipeIngredient(
                recipe_id=recipe.id,
                ingredient_id=item.ingredient_id,
                name=item.name.strip(),
                quantity=item.quantity,
                unit=item.unit,
                note=item.note,
                position=item.position if item.position is not None else index,
            )
        )
    for index, item in enumerate(payload.steps):
        session.add(
            RecipeStep(
                recipe_id=recipe.id,
                text=item.text.strip(),
                position=item.position if item.position is not None else index,
            )
        )


def _apply(recipe: Recipe, payload: RecipeWrite) -> None:
    recipe.title = payload.title.strip()
    recipe.description = payload.description
    recipe.category_id = payload.category_id
    recipe.favorite = payload.favorite
    recipe.rating = payload.rating
    recipe.user_note = payload.user_note
    recipe.servings = payload.servings
    recipe.prep_minutes = payload.prep_minutes
    recipe.cook_minutes = payload.cook_minutes
    recipe.image_ref = payload.image_ref
    recipe.updated_at = utcnow()


@router.get("", response_model=list[RecipeSummary])
def list_recipes(
    q: str | None = Query(default=None),
    favorite: bool | None = Query(default=None),
    category_id: str | None = Query(default=None),
    session: Session = Depends(get_session),
) -> list[RecipeSummary]:
    rows = session.exec(select(Recipe).order_by(Recipe.updated_at.desc())).all()
    needle = q.strip().lower() if q else None
    filtered = []
    for recipe in rows:
        if favorite is not None and recipe.favorite != favorite:
            continue
        if category_id and recipe.category_id != category_id:
            continue
        haystack = f"{recipe.title} {recipe.description or ''}".lower()
        if needle and needle not in haystack:
            continue
        filtered.append(_summary(session, recipe))
    return filtered


@router.post("", response_model=RecipeRead)
def create_recipe(
    payload: RecipeWrite, session: Session = Depends(get_session)
) -> RecipeRead:
    _validate_category(session, payload.category_id)
    recipe = Recipe(title=payload.title.strip())
    _apply(recipe, payload)
    session.add(recipe)
    session.commit()
    session.refresh(recipe)
    _replace_children(session, recipe, payload)
    session.commit()
    session.refresh(recipe)
    return _read(session, recipe)


@router.get("/{recipe_id}", response_model=RecipeRead)
def get_recipe(recipe_id: str, session: Session = Depends(get_session)) -> RecipeRead:
    recipe = session.get(Recipe, recipe_id)
    if not recipe:
        raise HTTPException(status_code=404, detail="Recipe not found")
    return _read(session, recipe)


@router.put("/{recipe_id}", response_model=RecipeRead)
def update_recipe(
    recipe_id: str, payload: RecipeWrite, session: Session = Depends(get_session)
) -> RecipeRead:
    _validate_category(session, payload.category_id)
    recipe = session.get(Recipe, recipe_id)
    if not recipe:
        raise HTTPException(status_code=404, detail="Recipe not found")
    _apply(recipe, payload)
    _replace_children(session, recipe, payload)
    session.add(recipe)
    session.commit()
    session.refresh(recipe)
    return _read(session, recipe)


@router.delete("/{recipe_id}")
def delete_recipe(recipe_id: str, session: Session = Depends(get_session)) -> dict[str, str]:
    recipe = session.get(Recipe, recipe_id)
    if not recipe:
        raise HTTPException(status_code=404, detail="Recipe not found")
    session.exec(delete(RecipeIngredient).where(RecipeIngredient.recipe_id == recipe.id))
    session.exec(delete(RecipeStep).where(RecipeStep.recipe_id == recipe.id))
    session.delete(recipe)
    session.commit()
    return {"status": "deleted"}
